clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java

```diff
@@ -417,6 +421,44 @@ public void run() {
         t.start();
     }
 
+    @Test
+    public void testSplitAwaitFlushComplete() throws Exception {
+        RecordAccumulator accum = createTestRecordAccumulator(1024, 10 * 1024, CompressionType.GZIP, 10);
+
+        // Create a big batch
+        byte[] value = new byte[256];
+        accum.append(new TopicPartition(topic, 0), 0L, null, value, null, null, maxBlockTimeMs);
+        accum.append(new TopicPartition(topic, 0), 0L, null, value, null, null, maxBlockTimeMs);
+        CountDownLatch flushInProgress = new CountDownLatch(1);
+        Iterator<ProducerBatch> incompleteBatches = accum.incompleteBatches().copyAll().iterator();
+        Assert.assertTrue(incompleteBatches.hasNext());
+        ProducerBatch producerBatch = incompleteBatches.next();
+        Assert.assertFalse(incompleteBatches.hasNext());
+
+        AtomicBoolean timedOut = new AtomicBoolean(false);
+
+        Thread thread = new Thread(() -> {
+            Assert.assertTrue(accum.hasIncomplete());
+            accum.beginFlush();
+            Assert.assertTrue(accum.flushInProgress());
+            try {
+                flushInProgress.countDown();
+                accum.awaitFlushCompletion(2000);
+            } catch (TimeoutException timeoutException) {
+              // Catch it and set the timedout variable
+              // This is the only valid path for this thread.
+                timedOut.set(true);
+            } catch (InterruptedException e) {
+            }
+        });
+        thread.start();
+        flushInProgress.await();
+        producerBatch.done(-1, -1, new RecordBatchTooLargeException());
+        producerBatch.produceFuture.done();
```

Comment:
It appears to me that the code in lines 456-457 will cause awaitFlushCompletion() to retry due to RecordBatchTooLargeException. However, this test does neither enqueue smaller batches nor incomplete batches. Thus, when awaitFlushCompletion() retries it will get producerBatch() again.

You used to call splitAndReenqueue(). I think that's what you should do here. Before calling producerBatch.producerFuture.done(), call deallocate(producerBatch) to remove it from incomplete batches first, and then call splitAndReenqueue() to add smaller batches back?

Reply:
I am calling 
- splitAndReenqueue
- produceBatch.done()
- produceBatch.produceFuture.done()
- accum.deallocate() 

in order. I think this emulates the exact behavior as in Sender.java class.